### ISSUE TITLE
Running STARsolo with Velocyto and building BAMs with gx/gn settings causes a segfault

### ISSUE THREAD

**Issue by detrout** [alexdobin/STAR#1898]
Hi,

I learned there was a Velocyto mode and tried adding it to my current command line

```
  --genomeDir ~/proj/genome/GRCm39-M32-male/ \
  --readFilesIn  next1/B01_13A_R2.fastq.gz,next2/B01_13A_R2.fastq.gz \
                         next1/B01_13A_R1.fastq.gz,next2/B01_13A_R1.fastq.gz \
  --readFilesCommand zcat   \
  --runThreadN 12   \
  --genomeLoad NoSharedMemory   \
  --outFilterMultimapNmax 20   \
  --alignSJoverhangMin 8   \
  --alignSJDBoverhangMin 1   \
  --outFilterMismatchNmax 999   \
  --outFilterMismatchNoverReadLmax 0.04   \
  --alignIntronMin 20   \
  --alignIntronMax 1000000  \
  --alignMatesGapMax 1000000   \
  --outSAMheaderHD @HD VN:1.4 SO:coordinate   \
  --outSAMunmapped Within   \
  --outFilterType BySJout   \
  --outSAMattributes NH HI AS NM MD CB CR CY UB UR UY gx gn   \
  --outSAMstrandField intronMotif   \
  --outSAMtype BAM SortedByCoordinate   \
  --sjdbScore 1  \
  --clipAdapterType CellRanger4   \
  --clip5pAdapterSeq AAGCAGTGGTATCAACGCAGAGTGAATGGG  \
   --outFilterScoreMin 30  \
   --soloUMIdedup 1MM_CR \
   --soloUMIfiltering MultiGeneUMI_CR  \
   --soloType CB_UMI_Complex  \
   --soloCellFilter EmptyDrops_CR \
   --soloCBmatchWLtype EditDist_2 \
   --soloCBposition 0_10_0_18 0_48_0_56 0_78_0_86  \
   --soloUMIposition 0_0_0_10 \
   --soloCBwhitelist inclusion/barcode-23_onlist.txt \
                                 inclusion/barcode-23_onlist.txt \
                                 inclusion/barcode-1_onlist_v2.txt  \
   --soloStrand Forward \
   --soloFeatures Gene GeneFull_Ex50pAS SJ Velocyto \
   --soloMultiMappers Unique  \
   --limitBAMsortRAM 68719476736 \
    --outTmpDir 13A_hand_velocyto/_STARtmp/ \
    --outFileNamePrefix 13A_hand_velocyto/
```

For some reason the container readAnnot.annotFeatures[P.pSolo.samAttrFeature].fAlign is empty with that mix of settings and the input files I'm currently using.  I was able to do an alignment & quantification with `--soloFeatures GeneFull_Ex50pAS SJ` 

I tried the following patch and it got most of the way through alignment but then ran into another crash on line 12 of SoloFeature_addBAMtags.cpp.

I don't know if this is a good way of solving this problem or if there's some precondition that isn't being satisfied.

```c++
@@ -409,8 +409,9 @@ int ReadAlign::alignBAM(Transcript const &trOut, uint nTrOut, uint iTrOut, uint
                     case ATTR_gx:
                         {
                             string algenes;
-                            for (auto &gg: readAnnot.annotFeatures[P.pSolo.samAttrFeature].fAlign[iTrOut])
-                                algenes += chunkTr->geID[gg] + ";";
+                            if (readAnnot.annotFeatures[P.pSolo.samAttrFeature].fAlign.size() > 0)
+                                for (auto &gg: readAnnot.annotFeatures[P.pSolo.samAttrFeature].fAlign[iTrOut])
+                                    algenes += chunkTr->geID[gg] + ";";
                             if (algenes.size()>0) {
                                 algenes.pop_back();
                             } else {
@@ -423,8 +424,9 @@ int ReadAlign::alignBAM(Transcript const &trOut, uint nTrOut, uint iTrOut, uint
                     case ATTR_gn:
                         {
                             string algenes;
-                            for (auto &gg: readAnnot.annotFeatures[P.pSolo.samAttrFeature].fAlign[iTrOut])
-                                algenes += chunkTr->geName[gg] + ";";
+                            if (readAnnot.annotFeatures[P.pSolo.samAttrFeature].fAlign.size() > 0)
+                                for (auto &gg: readAnnot.annotFeatures[P.pSolo.samAttrFeature].fAlign[iTrOut])
+                                    algenes += chunkTr->geName[gg] + ";";
                             if (algenes.size()>0) {
                                 algenes.pop_back();

```

Current try at the addBAMtags patch:

```c++
@@ -10,10 +10,10 @@ void SoloFeature::addBAMtags(char *&bam0, uint32 &size0, char *bam1)
 
     string cb="-";
     string umi="-";
-    if (readInfo[iread].cb+1 != 0)             
+    if (readInfo.size() > 0 && readInfo[iread].cb+1 != 0)
         cb  = pSolo.cbWLstr[readInfo[iread].cb];
     
-    if (readInfo[iread].umi+1 != 0)
+    if (readInfo.size() > 0 && readInfo[iread].umi+1 != 0)
         umi = convertNuclInt64toString(readInfo[iread].umi, pSolo.umiL);
     
     memcpy(bam1, bam0, size0);
```

Will try to write back after my next debug run finishes.

(2.7.10a does also still crash even when I get the read order correct)

**Comment by detrout** [comment]
There is a reasonable chance this is fixed in STAR 2.7.10b

I noticed there was an update, and am currently running against it.

**Comment by detrout** [comment]
Looks it was fixed in 2.7.10b.
